### PR TITLE
Fix race condition in P2P's disk buffer

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #xxxx

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

TODO:
- [x] Avoid write starvation
- [x] Test `ReadWriteLock`

### CODE REVIEW

**Comment by hendrikmakait** [comment]
cc @crusaderky. This should get rid of most (all?) windows-specific CI failures that were caused by a race between threads reading data from the buffer files and the buffer removing its directory on close.

**Comment by hendrikmakait** [comment]
Update: This should address

* `test_closed_worker_during_unpack`
* `test_restarting_during_unpack_raises_killed_worker`

both of which raise FS-related errors.

**Comment by fjetter** [comment]
I'm mildly concerned that we need a lock. Are we already offloading disk read/write or where is this coming from? I thought this was all still running on the event loop, i.e. same thread.

**Comment by hendrikmakait** [comment]
> I'm mildly concerned that we need a lock. Are we already offloading disk read/write or where is this coming from?

Reading is offloaded, which may cause a race between reads and cleanup that raises on Windows. The reason to offload is the shift to interleaving IO read and deserialization.

**Comment by hendrikmakait** [comment]
Note that the read-write lock is used as a shared-exclusive lock on the _directory_, it does not coordinate between actual IO reads and writes. I've just given in to the more typical naming convention.

File: distributed/shuffle/_disk.py [hunk 1]
```diff
@@ -71,11 +152,14 @@ async def _process(self, id: str, shards: list[bytes]) -> None:
         with log_errors():
             # Consider boosting total_size a bit here to account for duplication
             with self.time("write"):
-                with open(
-                    self.directory / str(id), mode="ab", buffering=100_000_000
-                ) as f:
-                    for shard in shards:
-                        f.write(shard)
+                with self._directory_lock.read():
```
Comment (fjetter) [inline distributed/shuffle/_disk.py:155]: If this is correct, this needs an inline comment
```python
with lock.read():
    file.write()
```

**Review by fjetter — COMMENTED** [review]
my above comment about possibly mixing up read/write should be addressed. Otherwise, this is good to go

File: distributed/shuffle/_disk.py (hunk 1 above)
Comment (fjetter) [reply distributed/shuffle/_disk.py:157]: I leave naming to you. I'm fine with the comment.

**Review by fjetter — APPROVED** [review]

**Pull request merged**